### PR TITLE
Updates on handling recorders

### PR DESCRIPTION
**Pull request by dericed**

### CODE REVIEW

**Comment by dericed** [comment]
This mixes a few updates, but generally removed bmdtools as an option (soon we can avoid triple-piped recording scenarios), and it swaps most of the use avfctl to dvrescue, and responds to a few feature requests.

**Review by libbyhopfauf — APPROVED** [review]
This seems to be working on my computers! I ended up having to go into recovery mode and resetting the permissions again (using the troubleshooting instructions here.) Now both devices show up on my computer via `avfctl -list_devices` and the vrecord GUI. Capture and packaging to MKV once capture is ended are working properly on my computers with this version.

```
libbyshopfauf@MacBook-Pro-55 ~ % avfctl -list_devices
2023-11-06 10:30:00.622 avfctl[1220:12548] Devices:
2023-11-06 10:30:00.623 avfctl[1220:12548] [0 - 0x8004601025e3c58] DV-VCR (Sony DSR-25)
2023-11-06 10:30:00.623 avfctl[1220:12548] [1 - 0x8004602010e007c] DSR-1500A (Sony DSR-1500A)
```
Screenshot 2023-11-06 at 10 35 11 AM



Computers tested:

- Macbook Pro 2019 running Ventura 13.5.1
- Macbook Pro 2021 running Ventura 13.5.1

Prior to resetting the permissions in recovery mode, I tried all of the other steps listed below (just in case it is helpful at all):

- In the vrecord GUI, I tried rescanning, refreshing vrecord and restarting my computer multiple times. If I selected "deck control"" in the GUI and click "repack", I get a message that the deck is stopped (similar to the example in https://github.com/amiaopensource/vrecord/issues/769):
Screenshot 2023-11-06 at 9 38 27 AM



- I tried uninstalling and reinstalling the Blackmagic drivers (including having to reset the permissions again.)
- Regarding my computers reading FireWire devices in general: if I ran `avfctl -list_devices`, I kept getting "`No devices found`." But if I ran `ioreg -l`, both DV devices connected are listed:

```
IOFireWireDevice@8004602010e007c  <clas$
    | |   |       |                   | {
    | |   |       |                   |   "FireWire Speed" = 0
    | |   |       |                   |   "FireWire Self IDs" = <816c08c0>
    | |   |       |                   |   "Vendor_ID" = 524358
    | |   |       |                   |   "FireWire Product Name" = "DSR-1500A"
    | |   |       |                   |   "FireWire Vendor Name" = "Sony"
    | |   |       |                   |   "FireWire Node ID" = 65473
    | |   |       |                   |   "FireWire Device ROM" = {"Offset 0"=<$
    | |   |       |                   |   "IOCFPlugInTypes" = {"CDCFCA94-F197-1$
    | |   |       |                   |   "GUID" = 576537726724997244
    | |   |       |                   |   "IOUserClientClass" = "IOFireWireUser$
    | |   |       |                   |   "Model_ID" = 3670189
	
	
    IOFireWireDevice@8004601025e3c58  <clas$
       | |   |           |               | {
       | |   |           |               |   "FireWire Speed" = 0
       | |   |           |               |   "FireWire Self IDs" = <816c08c2>
       | |   |           |               |   "Vendor_ID" = 524358
       | |   |           |               |   "FireWire Product Name" = "DSR-25"
       | |   |           |               |   "FireWire Vendor Name" = "Sony"
       | |   |           |               |   "FireWire Node ID" = 65535
       | |   |           |               |   "FireWire Device ROM" = {"Offset 0"=<$
       | |   |           |               |   "IOCFPlugInTypes" = {"CDCFCA94-F197-1$
       | |   |           |               |   "GUID" = 576537722452065368
       | |   |           |               |   "IOUserClientClass" = "IOFireWireUser$
       | |   |           |               |   "Model_ID" = 2097167
```

**Comment by tomackroyd** [comment]
Looking forward to trialling this. @libbyhopfauf where are the troubleshooting instructions you mentioned?

File: vrecord [hunk 1]
```diff
@@ -2007,6 +2014,7 @@ _review_all_options(){
     fi
     _review_option -n -d "digital" "WAVEFORM_SCALE_CHOICE" "${WAVEFORM_SCALE_OPTIONS[@]}"
     _review_option -n -d "Broadcast Range" "SIGNAL_VIEW_CHOICE" "${SIGNAL_OPTIONS[@]}"
+    _review_option -n -d "As is" "SIGNAL_INT_CHOICE" "${INT_OPTIONS[@]}"
```
Comment (privatezero) [inline vrecord:2017]: This line breaks things for me as `_review_option` will error by saying "As is" is not a valid option and exit. Does this work on your end?

File: vrecord [hunk 2]
```diff
@@ -554,17 +554,17 @@ _get_avfoundation_inputs(){
...
 
-_get_avfctl_inputs(){
-    # set avfctl input options
-    unset AVFCTL_DEVICES
-    while read avfctl_device ; do
-        AVFCTL_DEVICES+=("${avfctl_device}")
+_get_dvrescue_inputs(){
+    # set dvrescue input options
+    unset DVRESCUE_DEVICES
+    while read dvrescue_device ; do
```
Comment (privatezero) [inline vrecord:560]: I think this needs to be moved into an `else` clause so it doesn't run if the system is Linux (at least until I have time to make changes to the code to switch Linux side over to dvrescue as well. As written this displays the device name detected via dvrescue in the options (yay! that's ultimately a good thing) but is misleading to users until dvrescue is integrated as currently linux vrecord can only capture from the default DV device

**Review by privatezero — CHANGES_REQUESTED** [review]
Noticed a couple of gremlins!

File: vrecord [hunk 3]
```diff
@@ -2046,17 +1997,14 @@ _review_all_options(){
...
     if [[ "${DEVICE_INPUT_CHOICE}" = "0" ]] ; then
-        _review_option -n -d "ffmpeg" "DECKLINK_UTILITY_CHOICE" "${DECKLINK_UTILITIES[@]}"
         _review_option "VIDEO_INPUT_CHOICE" "${VIDEO_INPUT_OPTIONS[@]}"
         _review_option "AUDIO_INPUT_CHOICE" "${AUDIO_INPUT_OPTIONS[@]}"
         _review_option "VIDEO_BIT_DEPTH_CHOICE" "${VIDEO_BITDEPTH_OPTIONS[@]}"
         _review_option "AUDIO_MAPPING_CHOICE" "${CHANNEL_MAPPING_OPTIONS[@]}"
-        if [[ -z "${ALT_INPUT}" ]] && [[ "${DECKLINK_UTILITY_CHOICE}" == "ffmpeg" ]] ; then
-            _review_option "TIMECODE_CHOICE" "${TIMECODE_OPTIONS[@]}"
-        fi
+        _review_option "TIMECODE_CHOICE" "${TIMECODE_OPTIONS[@]}"
```
Comment (privatezero) [inline vrecord:2004]: this seems to break using most files for test inputs - suggest something like

```
if [[ -z "${ALT_INPUT}" ]] ; then
    _review_option "TIMECODE_CHOICE" "${TIMECODE_OPTIONS[@]}"
fi
```

File: vrecord (hunk 2 above)
Comment (privatezero) [reply vrecord:560]: changing order to something like the following seems to work on linux for now:

```
_get_dvrescue_inputs(){
    # set dvrescue input options
    if [[ "${OS_TYPE}" = "linux" ]] ; then
        DVRESCUE_DEVICES+=('Default DV Device')
    else
        unset DVRESCUE_DEVICES
        while read dvrescue_device ; do
            DVRESCUE_DEVICES+=("${dvrescue_device}")
        done < <(dvrescue -list_devices | grep "\[DV\]")
        if [[ "${#DVRESCUE_DEVICES[@]}" = 1 ]] ; then # default to first input if only one
            DVRESCUE_INPUT_CHOICE="${DVRESCUE_DEVICES[0]}"
        fi
    fi
}
```

File: vrecord (hunk 1 above)
Comment (privatezero) [reply vrecord:2017]: changing `"As is"` to "auto" fixes this on my end - this settings for this option do not, however, persist currently and are not apparently stored in the config file - was this intentional?

File: vrecord (hunk 2 above)
Comment (dericed) [reply vrecord:568]: The latest update does a lot more with getting linux to grab the device id and use it. Can you retest.

File: vrecord (hunk 3 above)
Comment (dericed) [reply vrecord:2020]: done

File: vrecord (hunk 1 above)
Comment (dericed) [reply vrecord:2033]: I switched to 'auto'

**Comment by dericed** [comment]
This is ready for a new review. Now avfctl is mostly gone and now dvrescue capture and playthrough is all done with just dvrescue and ffplay (or mpv) and not ffmpeg.

**Comment by dericed** [comment]
ping to @privatezero as https://github.com/mipops/dvrescue/issues/827#issuecomment-1964245329 relates to this.

**Comment by privatezero** [comment]
@dericed I've updated my issue there with results of another test

**Comment by privatezero** [comment]
@dericed I've made a modified version of this branch here: https://github.com/amiaopensource/vrecord/tree/updates-on-handling-recorders-linux

This branch adds a setting for allowing user to select capture tool for dv - this would be dvrescue on macs and dvrescue _or_ ffmpeg via iec61883 device on Linux. Any thoughts? Having a bit of choice/redundancy on the Linux side is a positive I think!

**Review by privatezero — APPROVED** [review]

**Comment by ablwr** [comment]
You did it

**Pull request merged**